edb/schema/constraints.py

```diff
@@ -372,27 +468,43 @@ def _constraint_args_from_ast(cls, schema, astnode, context):
 
         return args
 
-    def compile_expr_field(self, schema, context, field, value):
+    def compile_expr_field(
+        self,
+        schema: s_schema.Schema,
+        context: sd.CommandContext,
+        field: so.Field[Any],
+        value: s_expr.Expression,
+    ) -> s_expr.Expression:
         from edb.edgeql import compiler as qlcompiler
 
         if field.name in ('expr', 'subjectexpr'):
             if isinstance(self, CreateConstraint):
                 params = self._get_params(schema, context)
             else:
+                # TODO: this line is not hit in any test, is the assertion
+                # correct?
+                assert isinstance(
+                    self.scls,
+                    referencing.ReferencedInheritingObjectCommand
+                )
                 params = self.scls.get_params(schema)
             anchors, _ = (
                 qlcompiler.get_param_anchors_for_callable(
                     params, schema, inlined_defaults=False)
             )
             referrer_ctx = self.get_referrer_context(context)
+
             if referrer_ctx is not None:
+                assert isinstance(referrer_ctx.op, sd.ObjectCommand)
                 anchors['__subject__'] = referrer_ctx.op.scls
 
+            # TODO: it is true, the Parameter does not inherit Object,
+            # how to resolve the type: ignore below?
             return s_expr.Expression.compiled(
                 value,
                 schema=schema,
                 modaliases=context.modaliases,
-                anchors=anchors,
+                anchors=anchors,  # type: ignore
```

Comment:
Without `type: ignore`, mypy says:

```
Argument "anchors" to "compiled" of "Expression" has incompatible type "Dict[str, Parameter]"; expected "Optional[Mapping[Union[str, Type[SpecialAnchor]], Object]]"mypy(error)
```

It is right, because in the actual `Mapping[str, Parameter]`, Parameter does not inherit `Object`.
I am not sure what is the best way to proceed here.

Reply:
`Parameter` *does* inherit from `Object` (via `ObjectFragment`). I'm not sure why mypy complains here.

Reply:
I double checked, and this is the `Parameter` class from `ir` module, which doesn't inherit Object (specified in the return type of `compiler.get_param_anchors_for_callable`). The `Parameter` class that is an `ObjectFragment` is defined in `functions` module under schema.

Reply:
To add something more:

1. the dictionary returned by `qlcompiler.get_param_anchors_for_callable` is `Dict[str, irast.Parameter]`
2. irast.Parameter does not inherit from schema `Object`, hence the error report from mypy

Since this comment in `edb.ir.ast`:
```
        anchors:
            Predefined symbol table.  Maps identifiers
            (or ``qlast.SpecialAnchor`` instances) to specified
            schema objects or IR fragments.
```
says that `anchors` keys can be of type `str` or `qlast.SpecialAnchor`, and values can be "schema objects _or_ IR fragments", the place to correct is the type annotation for `anchors`:

instead of 
```py
    anchors: Optional[
        Mapping[Union[str, qlast.SpecialAnchorT], s_obj.Object]
    ] = None,
```

this:
```py
    anchors: Optional[
        Mapping[Union[str, qlast.SpecialAnchorT],
                Union[s_obj.Object, irast.Base]]
    ] = None,
```
**Note:** `stmtctx.py.init_context` already annotates `anchors` this way.

However, mypy still complains that `Dict[str, irast.Parameter] is not compatible with `Mapping[Union[str, qlast.SpecialAnchorT], Union[s_obj.Object, irast.Base]]`. IMHO, this looks surprising because `Mapping[str, irast.Base]` should match.

This happens because of the way union on keys and values are handled.

For example, the following two annotations are not equivalent:
```py
def example(value: Mapping[Union[str, int], Union[str, int]]):
    ...


def example2(value: Union[Mapping[str, int],
                          Mapping[int, str],
                          Mapping[str, str],
                          Mapping[int, int]]):
    ...
```

Reply:
@elprans, my conclusion is that `type: ignore` here is the best thing, because of these two things: https://github.com/python/mypy/issues/8477

Even changing the annotation for anchors everywhere to be proper (I gave a try to this), then we have issues in places where we would need to use the key as union of two possible key types.

Reply:
That's fine.